### PR TITLE
RL4J - Added update rule

### PR DESCRIPTION
**Pull request by aboulang2002**
## What changes were proposed in this pull request?

In this PR, the temporary class `DQNNeuralNetUpdateRule` is replaced by a generic one, `BaseUpdateRule`. Or more precisely, two: `GradientBasedUpdateRule` and `LabelBasedUpdateRule`, depending if the algorithm used generates _features-labels_ or _features-gradients_. Since both are valid choices, both are supported.

They both works like this: when a training batch is received, it is passed to a `IPolicyGradientAlgorithm` which returns a `Gradients` or a `UpdateLabels`. That result is then passed to a `INeuralNetUpdater` which abstracts how a Gradients/UpdateLabels is used to update the neural network.

I wrote a first version of this PR where Gradients/UpdateLabels were meant to be extended to support updates like ActorCritic. I.e. you would have, for example, a class extending a `BaseGradients` with a `getValueGradient()` and a `getPolicyGradient()`. This led to an explosion of generic parameters and, even if it was type-safer, I changed UpdateLabels and Gradients to have named labels and gradients.

You can use this to test:
```
package org.deeplearning4j.rl4j;

import org.datavec.api.transform.Operation;
import org.deeplearning4j.rl4j.agent.AgentLearner;
import org.deeplearning4j.rl4j.agent.learning.ILearningBehavior;
import org.deeplearning4j.rl4j.agent.learning.LearningBehavior;
import org.deeplearning4j.rl4j.agent.update.LabelBasedUpdateRule;
import org.deeplearning4j.rl4j.agent.update.policygradientalgorithm.IPolicyGradientAlgorithmFactory;
import org.deeplearning4j.rl4j.agent.update.policygradientalgorithm.dqn.DQNGradientAlgorithmFactory;
import org.deeplearning4j.rl4j.agent.update.neuralnetupdater.INeuralNetUpdaterFactory;
import org.deeplearning4j.rl4j.agent.update.neuralnetupdater.NeuralNetUpdaterFactory;
import org.deeplearning4j.rl4j.agent.update.neuralnetupdater.NeuralNetUpdater;
import org.deeplearning4j.rl4j.agent.update.policygradientalgorithm.dqn.IQLearningAlgorithm;
import org.deeplearning4j.rl4j.environment.Environment;
import org.deeplearning4j.rl4j.experience.ExperienceHandler;
import org.deeplearning4j.rl4j.experience.ReplayMemoryExperienceHandler;
import org.deeplearning4j.rl4j.learning.sync.Transition;
import org.deeplearning4j.rl4j.mdp.CartpoleEnvironment;
import org.deeplearning4j.rl4j.network.configuration.DQNDenseNetworkConfiguration;
import org.deeplearning4j.rl4j.network.dqn.DQNFactory;
import org.deeplearning4j.rl4j.network.dqn.DQNFactoryStdDense;
import org.deeplearning4j.rl4j.network.dqn.IDQN;
import org.deeplearning4j.rl4j.network.dqn.QOutput;
import org.deeplearning4j.rl4j.observation.transform.TransformProcess;
import org.deeplearning4j.rl4j.policy.DQNPolicy;
import org.deeplearning4j.rl4j.policy.EpsGreedy;
import org.deeplearning4j.rl4j.policy.INeuralNetPolicy;
import org.deeplearning4j.rl4j.policy.IPolicy;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.factory.Nd4j;
import org.nd4j.linalg.learning.config.Adam;

public class AgentLearnerCartpole {

    public static DQNFactoryStdDense.Configuration CARTPOLE_NET =
            DQNFactoryStdDense.Configuration.builder()
                    .l2(0.001).updater(new Adam(0.0005)).numHiddenNodes(16).numLayer(3).build();

    public static void main(String[] args) {
        // Build the network
        DQNDenseNetworkConfiguration netConf = DQNDenseNetworkConfiguration.builder()
                .numLayers(3)
                .numHiddenNodes(16)
                .l2(0.001)
                .updater(new Adam(0.0005))
                .build();
        DQNFactory factory = new DQNFactoryStdDense(netConf);
        IDQN network = factory.buildDQN(new int[] { 4 }, 2);

        // Build the environment and the TransformProcess
        Environment<Integer> cartpoleEnv = new CartpoleEnvironment();
        TransformProcess transformProcess = TransformProcess.builder()
                .transform("data", new ArrayToINDArrayTransform())
                .build("data");

        // Build the learning behavior
        ExperienceHandler<Integer, Transition<Integer>> experienceHandler = new ReplayMemoryExperienceHandler(10000, 64, Nd4j.getRandom());
        IPolicyGradientAlgorithmFactory<IQLearningAlgorithm<Integer>, QOutput> gradientAlgorithmFactory = DQNGradientAlgorithmFactory.builder()
                .isDoubleDQN(true);
        INeuralNetUpdaterFactory<QOutput> nnUpdaterFactory = NeuralNetUpdaterFactory.builder(NeuralNetUpdater.LabelsToGradientsSource.UseCurrentToComputeGradients)
                .targetUpdateFrequency(50);
        LabelBasedUpdateRule updateRule = new LabelBasedUpdateRule(network, gradientAlgorithmFactory, nnUpdaterFactory);
        ILearningBehavior<Integer> learningBehavior = LearningBehavior.<Integer, Transition<Integer>>builder()
                .experienceHandler(experienceHandler)
                .updateRule(updateRule)
                .build();

        // Build the policies -- a greedy policy wrapped around an epsilon-greedy one
        INeuralNetPolicy<Integer> policy = new DQNPolicy<Integer>((IDQN)updateRule.getCurrentNetwork());
        IPolicy<Integer> epsGreedyPolicy = EpsGreedy.<Integer>builder()
                .policy(policy)
                .actionSchema(cartpoleEnv.getSchema().getActionSchema())
                .minEpsilon(0.1)
                .epsilonNbStep(3000)
                .rnd(Nd4j.getRandomFactory().getNewRandomInstance(123))
                .build();

        // Build the AgentLearner
        AgentLearner<Integer> agentLearner = AgentLearner.builder(cartpoleEnv, transformProcess, epsGreedyPolicy, learningBehavior)
                .maxEpisodeSteps(200)
                .build();

        // Train for 100 episodes
        for(int i = 0; i < 100; ++i) {
            agentLearner.run();
            System.out.println(String.format("%d : %f", i, agentLearner.getReward()));
        }
    }

    public static class ArrayToINDArrayTransform implements Operation<double[], INDArray> {
        @Override
        public INDArray transform(double[] data) {
            return Nd4j.create(data);
        }
    }
}
```
## How was this patch tested?

Manual tests and unit tests. 

## Quick checklist

The following checklist helps ensure your PR is complete:

- [X] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.org/eclipse-contributors) page for details
- [X] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [X] Created tests for any significant new code additions.
- [X] Relevant tests for your changes are passing.
- [X] Added javadoc

### CODE REVIEW

**Comment by aboulang2002** [comment]
There is a small problem with the code in the description above. The policy is created with `network` which is used as the _target_ network in the update rule. I will make sure everything is set correctly in the next PR.

**Comment by saudet** [comment]
The Cartpole example doesn't converge very well for me anymore with these changes.
We'll need to figure out what broke and fix it before we can merge this...

**Comment by aboulang2002** [comment]
I am closing this PR. I will divide the changes in smaller PRs instead, that will be easier.
